### ISSUE TITLE
Strange clippy suggestion for `clone_from`

### ISSUE THREAD

**Issue by cavivie** [rust-lang/rust-clippy#12929]
### Description

```rust
use std::{fs::File, io::{BufReader, BufRead}};

fn main() -> std::io::Result<()> {
    let file = File::open("/dev/random")?;
    let mut rand_buf = BufReader::new(file);
    let mut rand_utf8_str = String::new();

    while rand_buf.read_line(&mut rand_utf8_str).is_err() {}
    if let Some(val) = rand_utf8_str.strip_suffix('\n') { // immutable borrow occurs here
        // clippy suggest me to change it
        val.clone_into(&mut rand_utf8_str); // can not compile, why is this suggested?
        // rand_utf8_str = val.to_owned(); // it is the original writing, can compile
    }

    println!("random utf8 string: {:?}", rand_utf8_str);

    Ok(())
 }
```

```shell
cargo clippy -- -D warnings
```

### Version

```text
rustc 1.79.0 (129f3b996 2024-06-10)
binary: rustc
commit-hash: 129f3b9964af4d4a709d1383930ade12dfe7c081
commit-date: 2024-06-10
host: x86_64-unknown-linux-gnu
release: 1.79.0
LLVM version: 18.1.7
```

**Comment by y21** [comment]
This largely looks like a duplicate of #12460 (at least the pattern is very similar), which itself is also a duplicate. All of the borrowck error related FPs will be fixed by #12756.

**Comment by cavivie** [comment]
dup
